session/session.go

```diff
@@ -0,0 +1,135 @@
+package session
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"launchpad.net/gozk"
+)
+
+type ZKSessionEvent uint
+
+var ErrZKSessionNotConnected = errors.New("unable to connect to ZooKeeper")
+
+const (
+	SessionClosed ZKSessionEvent = iota
+	SessionDisconnected
+	SessionReconnected
+	SessionFailed
+)
+
+type ZKSession struct {
+	servers     string
+	recvTimeout time.Duration
+	conn        *zookeeper.Conn
+	clientID    *zookeeper.ClientId
+	events      <-chan zookeeper.Event
+	sEvents     chan ZKSessionEvent
+	mu          sync.Mutex
+}
+
+func NewZKSession(servers string, recvTimeout time.Duration) (*ZKSession, <-chan ZKSessionEvent, error) {
+	conn, events, err := zookeeper.Dial(servers, recvTimeout)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	s := &ZKSession{
+		servers:     servers,
+		recvTimeout: recvTimeout,
+		conn:        conn,
+		clientID:    conn.ClientId(),
+		events:      events,
+		sEvents:     make(chan ZKSessionEvent),
+	}
+
+	event := <-events
+	if event.State != zookeeper.STATE_CONNECTED {
+		return nil, nil, ErrZKSessionNotConnected
+	}
+
+	go s.manage()
+
+	return s, s.sEvents, nil
+}
+
+func (s *ZKSession) manage() {
+	for {
```

Comment:
> This can just be:

Actually, this won't work in this case because on redial I swap out the events channel. With the example you provided it won't pick up the new channel but will continue to range over the old channel.